src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs

```diff
@@ -22,9 +22,29 @@ public static void CopyFile(string sourceFullPath, string destFullPath, bool ove
 
             // Copy the contents of the file from the source to the destination, creating the destination in the process
             using (var src = new FileStream(sourceFullPath, FileMode.Open, FileAccess.Read, FileShare.Read, DefaultBufferSize, FileOptions.None))
-            using (var dst = new FileStream(destFullPath, overwrite ? FileMode.Create : FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, DefaultBufferSize, FileOptions.None))
             {
-                Interop.CheckIo(Interop.Sys.CopyFile(src.SafeFileHandle, dst.SafeFileHandle));
+                int result = Interop.Sys.CopyFile(src.SafeFileHandle, sourceFullPath, destFullPath, overwrite ? 1 : 0);
+
+                if (result < 0)
+                {
+                    Interop.ErrorInfo error = Interop.Sys.GetLastErrorInfo();
+
+                    // If we fail to open the file due to a path not existing, we need to know whether to blame
+                    // the file itself or its directory.  If we're creating the file, then we blame the directory,
+                    // otherwise we blame the file.
+                    //
+                    // When opening, we need to align with Windows, which considers a missing path to be
+                    // FileNotFound only if the containing directory exists.
+
+                    bool isDirectory = (error.Error == Interop.Error.ENOENT) &&
+                        (overwrite || !DirectoryExists(Path.GetDirectoryName(Path.TrimEndingDirectorySeparator(destFullPath!))!));
```

Comment:
Why is the `!` needed on `destFullPath!`?

Reply:
For that matter why are either of the `!` suppressions needed?

Reply:
It should not be needed. I copied it from https://github.com/dotnet/corefx/blob/d3911035f2ba3eb5c44310342cc1d654e42aa316/src/Common/src/CoreLib/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs#L55. It should probably be removed from both places.